src/routes/threlte/projectUtils.ts

```diff
@@ -0,0 +1,544 @@
+	const tolerance = CIRCLE_TOLERANCE // in meters
+	const radius = start.distanceTo(center)
+	const k = tolerance / radius
+	// more precise but slower to calculate:
+	// const n = Math.ceil(Math.PI / Math.acos(1 - k))
+	// faster to calculate, at most only overestimates by 1:
+	let n = Math.ceil(Math.PI / Math.sqrt(2 * k))
+	const segmentAngle = (2 * Math.PI) / n
+	const segmentLength = radius * segmentAngle
+	if (clockwise) n = -n
+
+	const startAngle = Math.atan2(start.y - center.y, start.x - center.x)
+
+	const lineVertices = []
+	lineVertices.push(start.clone())
+	for (let i = 1; i <= Math.abs(n); i++) {
+		const theta = ((2 * Math.PI) / n) * i + startAngle
+		const xComponent = radius * Math.cos(theta)
+		const yComponent = radius * Math.sin(theta)
+		const point = new Vector2(xComponent, yComponent).add(center)
+		lineVertices.push(point)
+
+		const distanceToEnd = point.distanceTo(end)
+		if (distanceToEnd <= segmentLength) {
+			lineVertices.push(end.clone())
+			break
+		}
+	}
+	return lineVertices
+}
+
+export function circleToPoints(centerPoint: Vector2Like, radius: number): Vector2[] {
+	// this is 2D function
+
+	// see https://math.stackexchange.com/a/4132095/816177
+	const tolerance = CIRCLE_TOLERANCE // in meters
+	const k = tolerance / radius
+	// more precise but slower to calculate:
+	// const n = Math.ceil(Math.PI / Math.acos(1 - k))
+	// faster to calculate, at most only overestimates by 1:
+	const n = Math.ceil(Math.PI / Math.sqrt(2 * k))
+
+	const lineVertices = []
+	for (let i = 0; i <= n; i++) {
+		const theta = ((2 * Math.PI) / n) * i
+		const xComponent = radius * Math.cos(theta)
+		const yComponent = radius * Math.sin(theta)
+		const point = new Vector2(xComponent, yComponent).add(centerPoint)
+		lineVertices.push(point)
+	}
+	return lineVertices
+}
+
+export function promoteTo3(points: Vector2[]): Vector3[] {
+	const points3 = []
+	for (const point of points) {
+		points3.push(new Vector3(point.x, point.y, 0))
+	}
+	return points3
+}
+
+export function flatten(points: Vector3[]): number[] {
+	const pointsFlat = []
+	for (const point of points) {
+		pointsFlat.push(point.x, point.y, point.z)
+	}
+	return pointsFlat
+}
+
+function isStringInt(s: string, errorCallback: { (id: any): void; (arg0: string): void }): boolean {
+	if (typeof s !== "string") console.error("[proectUtils.ts] [isStringInt]", s, "is not a string:", typeof s)
+	const isInt = !Number.isNaN(parseInt(s, 10))
+	if (!isInt) errorCallback(s)
+	return isInt
+}
+
+function reduceToInts(data: string[], errorCallback: (id: any) => void): number[] {
+	function reducer(acc: number[], id: string): number[] {
+		return isStringInt(id, errorCallback) ? [...acc, parseInt(id, 10)] : acc
+	}
+	return data.reduce(reducer, [])
+}
+
+function notEmpty(array: unknown[]): boolean {
+	return (array && Array.isArray(array) && array.length > 0)
+}
+
+function checkWasmMessage(message: Message, abort = true, logError = true): boolean {
```

Comment:
love this!

Reply:
it is pretty sweet! all credit goes to ts-auto-guard for making the type guards easy to make! it doesn't generate fully correct code but it is mostly so, only takes a couple minutes to fix it up after regenerating with `pnpm type:guards`